### PR TITLE
Fix conflict between MessageFactory and MergeFrom

### PR DESCRIPTION
**Pull request by zunger-humu**
This fixes a bug in the Python code where reflection generates classes structurally identical to, and with almost the same names as, but not actually equal to, the protoc-generated classes. This means that MergeFrom between two seemingly identical classes fails because of a type mismatch.

This breaks one of the most important ways to use MessageFactory (reflection), where someone stores both the bytes of a protobuf and the name of the proto class, and later on reads it back using the same: the resulting object created is of one of the reflected types, and any other code trying to interoperate will be using a generated type, and Hilarity Ensues.

This PR is a resurrection of a previous one (https://github.com/google/protobuf/pull/3597) which fixed this issue in `reflection.MakeClass`; that PR was closed when MakeClass was deprecated in favor of MessageFactory, but the bug persists. This PR fixes the bug in MessageFactory.

### CODE REVIEW

**Comment by zunger-humu** [comment]
Fun note about this bug: Because MessageFactory is clever about caching, if you add the test case without also adding the fix, not only does that test case fail, but a bunch of other random test cases across protobuf start to fail for the same reason. :)

File: python/google/protobuf/internal/reflection_test.py [hunk 1]
```diff
@@ -1505,6 +1508,24 @@ def testMergeFromBug(self):
     message2.MergeFrom(message1)
     self.assertFalse(message2.HasField('optional_nested_message'))
 
+  def testReflectionDoesntBreakMerging(self):
```
Comment (anandolee) [inline python/google/protobuf/internal/reflection_test.py]: Add the test to message_factory_test instead.

File: src/google/protobuf/map_unittest.proto [hunk 2]
```diff
@@ -61,6 +61,11 @@ message TestMap {
   map<int32   , ForeignMessage> map_int32_foreign_message = 17;
   map<string  , ForeignMessage> map_string_foreign_message = 18;
   map<int32   , TestAllTypes> map_int32_all_types = 19;
+
```
Comment (anandolee) [inline src/google/protobuf/map_unittest.proto]: Are you able to reproduce the error by current .proto files? I guess google.protobuf.unittest_pb2.TestAllTypes should be good enough.

File: python/google/protobuf/message_factory.py [hunk 3]
```diff
@@ -122,6 +115,32 @@ def GetMessages(self, files):
         extended_class.RegisterExtension(extension)
     return result
 
+  def ClearCache(self):
```
Comment (anandolee) [inline python/google/protobuf/message_factory.py]: Please do not add new APIs for fixing PR. 

If you want to add new APIs, we should have a API review (as a public library, we are very careful to add new APIs). New API should be supported in both pure python and cpp extension. Also need to have test coverage for it.

File: python/google/protobuf/message_factory.py [hunk 4]
```diff
@@ -122,6 +115,32 @@ def GetMessages(self, files):
...
+    """Clear the MessageFactory's cache of which classes it knows about. This
+    is typically only something you need to do if you've been programmatically
+    importing _pb2.py files and want to make sure the factory understands them
+    correctly.
+    """
+    self._classes = {}
+
+  def _GetPrototypeClass(self, descriptor):
+    try:
```
Comment (anandolee) [inline python/google/protobuf/message_factory.py]: No need to add  _GetPrototypeClass(self, descriptor), just move the try/except to GetPrototype(self, descriptor) should be enough

File: python/google/protobuf/internal/reflection_test.py (hunk 1 above)
Comment (zunger-humu) [reply python/google/protobuf/internal/reflection_test.py]: Done.

File: python/google/protobuf/message_factory.py (hunk 3 above)
Comment (zunger-humu) [reply python/google/protobuf/message_factory.py]: Done.

File: python/google/protobuf/message_factory.py (hunk 4 above)
Comment (zunger-humu) [reply python/google/protobuf/message_factory.py]: Done.

File: src/google/protobuf/map_unittest.proto (hunk 2 above)
Comment (zunger-humu) [reply src/google/protobuf/map_unittest.proto]: Oops, good catch. The old PR needed this, the new one creates the bug just fine without it. :)

**Comment by anandolee** [comment]
Thanks so much for your fix @zunger-humu, and your change is correct.

However, I just checked our internal code that someone has added a similar _concrete_class check in GeneratedProtocolMessageType which can also make your test pass. The change was added in May, you'd expect the see the fix in next release.

**Comment by zunger-humu** [comment]
Sweet! But is this change in yet? I synced against git yesterday, and my
test definitely doesn't pass with this change commented out.

**Comment by anandolee** [comment]
Not, the change is not in github yet. It is in google's internal code base, which will be integrated to github when preparing next release.

The integrated RP for 3.6.0 for example:
https://github.com/google/protobuf/pull/4387

**Comment by anandolee** [comment]
Closing it as 3.7.0 is released

**Pull request closed without merging**